### PR TITLE
Add +T no-CTCP usermode

### PR DESCRIPTION
**Pull request by clukawski**
Adds no-CTCP usermode as per #1007

### CODE REVIEW

**Comment by slingamn** [comment]
This looks great. I'll leave it open a little longer before merging.

**Comment by slingamn** [comment]
@ajaspers points out that this should prevent *receiving* CTCP messages, not sending them.

**Comment by KoraggKnightWolf** [comment]
Sending CTCPs should indeed still be possible if the user has +T also what an optional oper-override option? Either just a config option or an oper priv? It might be useful for opers to know what client and version a user has (regarding the client parsing issue Insp 3 has revealed for example).

**Comment by slingamn** [comment]
I think I'm uncomfortable with this as an operator privilege. (It also presents a UI problem: normally we make operators exercise their privileges explicitly, e.g., with `SAMODE` or `SAJOIN`, but there's no `SAPRIVMSG`.)

File: irc/handlers.go [hunk 1]
```diff
@@ -1874,7 +1874,11 @@ func messageHandler(server *Server, client *Client, msg ircmsg.IrcMessage, rb *R
 		return false
 	}
 
-	if rb.session.isTor && utils.IsRestrictedCTCPMessage(message) {
+	isCTCP := utils.IsRestrictedCTCPMessage(message)
+	if isCTCP && client.modes.HasMode(modes.UserNoCTCP) {
```
Comment (slingamn) [inline irc/handlers.go:1878]: This is still checking the mode on the sender, not the recipient?

**Comment by slingamn** [comment]
This looks pretty good. Could you:

1. Remove `ERR_CANTSENDTOUSER`
2. Squash this to a single commit on top of master?

File: irc/handlers.go [hunk 2]
```diff
@@ -1938,6 +1938,12 @@ func dispatchMessageToTarget(client *Client, tags map[string]string, histType hi
 			}
 			return
 		}
+
+		// Restrict CTCP message for target user with +T
+		if message.IsRestrictedCTCPMessage() && user.modes.HasMode(modes.UserNoCTCP) {
```
Comment (slingamn) [inline irc/handlers.go:1943]: Optional performance-related feedback: either reverse the order of these checks, or cache the result of `IsRestrictedCTCPMessage` for use lower down with Tor sessions.

Comment (clukawski) [reply irc/handlers.go:1943]: :+1:

**Comment by clukawski** [comment]
Squashed and removed the unused numeric.

**Comment by clukawski** [comment]
Ok I reversed those checks as well, all squashed again

**Pull request merged**